geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/graph/TriangleCount.java

```diff
@@ -0,0 +1,115 @@
+import com.antgroup.geaflow.dsl.common.algo.AlgorithmUserFunction;
+import com.antgroup.geaflow.dsl.common.data.Row;
+import com.antgroup.geaflow.dsl.common.data.RowEdge;
+import com.antgroup.geaflow.dsl.common.data.RowVertex;
+import com.antgroup.geaflow.dsl.common.data.impl.ObjectRow;
+import com.antgroup.geaflow.dsl.common.function.Description;
+import com.antgroup.geaflow.dsl.common.types.StructType;
+import com.antgroup.geaflow.dsl.common.types.TableField;
+import com.antgroup.geaflow.model.graph.edge.EdgeDirection;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+@Description(name = "triangle_count", description = "built-in udga for Triangle Count.")
+public class TriangleCount implements AlgorithmUserFunction<Object, ObjectRow> {
+
+    private AlgorithmRuntimeContext<Object, ObjectRow> context;
+
+    private final int maxIteration = 2;
+
+    private String vertexType = null;
+
+    private final Set<Long> excludeSet = Sets.newHashSet();
+
+    @Override
+    public void init(AlgorithmRuntimeContext<Object, ObjectRow> context, Object[] params) {
+        this.context = context;
+        if (params.length >= 1) {
+            assert params[0] instanceof String : "Vertex type parameter should be string.";
+            vertexType = (String) params[0];
+        }
+        assert params.length <= 1 : "Maximum parameter limit exceeded.";
+    }
+
+    @Override
+    public void process(RowVertex vertex, Iterator<ObjectRow> messages) {
+        if (context.getCurrentIterationId() == 1L) {
+            if (Objects.nonNull(vertexType) && !vertexType.equals(vertex.getLabel())) {
+                excludeSet.add((Long) vertex.getId());
+                return;
+            }
+
+            List<RowEdge> rowEdges = context.loadEdges(EdgeDirection.BOTH);
+            List<Object> neighborInfo = Lists.newArrayList();
+            neighborInfo.add((long) rowEdges.size());
+            for (RowEdge rowEdge : rowEdges) {
+                neighborInfo.add(rowEdge.getTargetId());
+            }
+            ObjectRow msg = ObjectRow.create(neighborInfo.toArray());
+            for (int i = 1; i < neighborInfo.size(); i++) {
+                context.sendMessage(neighborInfo.get(i), msg);
+            }
+            context.sendMessage(vertex.getId(), ObjectRow.create(0L));
+            context.updateVertexValue(msg);
+        } else if (context.getCurrentIterationId() <= maxIteration) {
+            if (Objects.nonNull(vertexType) && !vertexType.equals(vertex.getLabel())) {
+                return;
+            }
+            long count = 0;
+            Set<Long> sourceSet = row2Set(vertex.getValue());
+            while (messages.hasNext()) {
+                ObjectRow msg = messages.next();
+                Set<Long> targetSet = row2Set(msg);
+                targetSet.retainAll(sourceSet);
```

Comment:
The vertex that receives the message needs to traverse all the messages to obtain the complete and correct set of vertices that meet the requirements.  Before this, the intersection vertices obtained are either in pending or have already appeared in previous messages.